File: app/src/main/java/com/xmx/tango/module/operate/TangoOperator.java
package com.xmx.tango.module.operate;

import com.xmx.tango.core.Constants;
import com.xmx.tango.common.data.DataManager;
import com.xmx.tango.module.calendar.DateData;
import com.xmx.tango.module.calendar.DateDataEntityManager;
import com.xmx.tango.module.tango.Tango;
import com.xmx.tango.module.tango.TangoConstants;
import com.xmx.tango.module.tango.TangoEntityManager;

import org.greenrobot.eventbus.EventBus;

import java.util.Date;

/**
 * Created by The_onE on 2016/10/7.
 */
public class TangoOperator {
    public int study;
    public int review;
    int todayConsecutive = 0;

    private Tango prevTango;
    private int prevStudy;
    private int prevReview;

    private static TangoOperator instance;

    public synchronized static TangoOperator getInstance() {
        if (null == instance) {
            instance = new TangoOperator();
        }
        return instance;
    }

    private TangoOperator() {
        study = DataManager.getInstance().getTangoStudy();
        review = DataManager.getInstance().getTangoReview();

        Date last = new Date(DataManager.getInstance().getResetLastTime());
        Date now = new Date();
        if (!Constants.isSameDate(now, last)) {
            if (last.getTime() > 0) {
                // 更新上次签到日数据
                DateData dateData = DateDataEntityManager.INSTANCE
                        .selectLatest("addTime", false,
                                "Year=" + (last.getYear() + 1900),
                                "Month=" + (last.getMonth() + 1),
                                "Date=" + last.getDate());
                if (dateData != null) {
                    DateDataEntityManager.INSTANCE.updateData(dateData.getId(),
                            "Study=" + study,
                            "Review=" + review);
                } else {
                    dateData = new DateData();
                    dateData.setYear(last.getYear() + 1900);
                    dateData.setMonth(last.getMonth() + 1);
                    dateData.setDate(last.getDate());
                    dateData.setCheckIn(1);
                    dateData.setStudy(study);
                    dateData.setReview(review);
                    dateData.setAddTime(now);
                    DateDataEntityManager.INSTANCE.insertData(dateData);
                }
            }
            // 今天打卡签到
            DateData todayData = DateDataEntityManager.INSTANCE
                    .selectLatest("addTime", false,
                            "Year=" + (now.getYear() + 1900),
                            "Month=" + (now.getMonth() + 1),
                            "Date=" + now.getDate());
            if (todayData == null) {
                todayData = new DateData();
                todayData.setYear(now.getYear() + 1900);
                todayData.setMonth(now.getMonth() + 1);
                todayData.setDate(now.getDate());
                todayData.setCheckIn(1);
                todayData.setAddTime(now);
                DateDataEntityManager.INSTANCE.insertData(todayData);
            }
            study = 0;
            DataManager.getInstance().setTangoStudy(0);
            review = 0;
            DataManager.getInstance().setTangoReview(0);

            DataManager.getInstance().setReviewFrequency(TangoConstants.INSTANCE.getREVIEW_FREQUENCY());

            DataManager.getInstance().setResetLastTime(now.getTime());
        }
    }

    public void remember(Tango tango) {
        if (tango != null && tango.getId() > 0) {
            prevTango = tango;
            prevStudy = study;
            prevReview = review;

            Date last = tango.getLastTime();
            Date now = new Date();
            int frequency = tango.getFrequency();
            int goal = DataManager.getInstance().getTangoGoal();
            if (!Constants.isSameDate(now, last)) {
                todayConsecutive = 0;
                if (last.getTime() > 0) { //复习
                    review++;
                    DataManager.getInstance().setTangoReview(review);
                    //DataManager.getInstance().setLastTime(now.getTime());
                    if (frequency > 0) {
                        frequency--;
                    }
                } else { //学习
                    study++;
                    frequency = TangoConstants.INSTANCE.getREVIEW_FREQUENCY();
                    DataManager.getInstance().setTangoStudy(study);
                }
            } else if (study >= goal) {
                todayConsecutive++;
                if (todayConsecutive > TangoConstants.INSTANCE.getTODAY_CONSECUTIVE_REVIEW_MAX()) {
                    todayConsecutive = 0;
                    int frequencyMax = DataManager.getInstance().getReviewFrequency();
                    frequencyMax--;
                    DataManager.getInstance().setReviewFrequency(frequencyMax);
                }
            }

            int score = TangoConstants.INSTANCE.getREMEMBER_SCORE() - (study + review) / TangoConstants.INSTANCE.getTIRED_COEFFICIENT();
            score = Math.max(score, TangoConstants.INSTANCE.getREMEMBER_MIN_SCORE());
            TangoEntityManager.INSTANCE.updateData(tango.getId(),
                    "Score=" + (tango.getScore() + score),
                    "Frequency=" + frequency,
                    "LastTime=" + new Date().getTime());
            EventBus.getDefault().post(new OperateTangoEvent());
        }
    }

    public void forget(Tango tango) {
        if (tango != null && tango.getId() > 0) {
            prevTango = tango;
            prevStudy = study;
            prevReview = review;

            Date last = tango.getLastTime();
            Date now = new Date();
            int frequency = tango.getFrequency();
            if (!Constants.isSameDate(now, last)) {
                if (last.getTime() > 0) { //复习
                    frequency++;
                    if (frequency > TangoConstants.INSTANCE.getREVIEW_FREQUENCY()) {
                        frequency = TangoConstants.INSTANCE.getREVIEW_FREQUENCY();
                    }
                }
            }

            TangoEntityManager.INSTANCE.updateData(tango.getId(),
                    "Score=" + (tango.getScore() + TangoConstants.INSTANCE.getFORGET_SCORE()),
                    "Frequency=" + frequency);
            //"LastTime=" + new Date().getTime());
            EventBus.getDefault().post(new OperateTangoEvent());
        }
    }

    public void rememberForever(Tango tango) {
        if (tango != null && tango.getId() > 0) {
            prevTango = tango;
            prevStudy = study;
            prevReview = review;

            Date last = tango.getLastTime();
            Date now = new Date();
            int frequency = tango.getFrequency();
            if (!Constants.isSameDate(now, last)) {
                if (last.getTime() > 0) { //复习
                    review++;
                    DataManager.getInstance().setTangoReview(review);
                    //DataManager.getInstance().setLastTime(now.getTime());
                } else { //学习
                    study++;
                    DataManager.getInstance().setTangoStudy(study);
                    //DataManager.getInstance().setLastTime(now.getTime());
                }
            }
            frequency = -1;

            TangoEntityManager.INSTANCE.updateData(tango.getId(),
                    "Score=" + (tango.getScore() + TangoConstants.INSTANCE.getREMEMBER_FOREVER_SCORE()),
                    "Frequency=" + frequency,
                    "LastTime=" + new Date().getTime());
            EventBus.getDefault().post(new OperateTangoEvent());
        }
    }

    public void rightWithoutHint(Tango tango) {
        if (tango != null && tango.getId() > 0) {
            prevTango = tango;
            prevStudy = study;
            prevReview = review;

            Date last = tango.getLastTime();
            Date now = new Date();
            int frequency = tango.getFrequency();
            if (!Constants.isSameDate(now, last)) {
                if (last.getTime() > 0) { //复习
                    review++;
                    DataManager.getInstance().setTangoReview(review);
                }
                if (frequency > 0) {
                    frequency -= 2;
                    if (frequency < 0) {
                        frequency = 0;
                    }
                }
            } else {
                todayConsecutive++;
                if (todayConsecutive > TangoConstants.INSTANCE.getTODAY_CONSECUTIVE_REVIEW_MAX()) {
                    todayConsecutive = 0;
                    int frequencyMax = DataManager.getInstance().getReviewFrequency();
                    frequencyMax--;
                    DataManager.getInstance().setReviewFrequency(frequencyMax);
                }
            }

            TangoEntityManager.INSTANCE.updateData(tango.getId(),
                    "Score=" + (tango.getScore() + TangoConstants.INSTANCE.getREMEMBER_SCORE() * 2),
                    "Frequency=" + frequency,
                    "LastTime=" + new Date().getTime());
            EventBus.getDefault().post(new OperateTangoEvent());
        }
    }

    public void rightWithHint(Tango tango) {
        if (tango != null && tango.getId() > 0) {
            prevTango = tango;
            prevStudy = study;
            prevReview = review;

            Date last = tango.getLastTime();
            Date now = new Date();
            int frequency = tango.getFrequency();
            if (!Constants.isSameDate(now, last)) {
                if (last.getTime() > 0) { //复习
                    review++;
                    DataManager.getInstance().setTangoReview(review);
                }
                if (frequency > 0) {
                    frequency--;
                }
            }

            TangoEntityManager.INSTANCE.updateData(tango.getId(),
                    "Score=" + (tango.getScore() + TangoConstants.INSTANCE.getREMEMBER_SCORE()),
                    "Frequency=" + frequency,
                    "LastTime=" + new Date().getTime());
            EventBus.getDefault().post(new OperateTangoEvent());
        }
    }

    public void wrong(Tango tango) {
        if (tango != null && tango.getId() > 0) {
            prevTango = tango;
            prevStudy = study;
            prevReview = review;

            Date last = tango.getLastTime();
            Date now = new Date();
            int frequency = tango.getFrequency();
            if (!Constants.isSameDate(now, last)) {
                if (last.getTime() > 0) { //复习
                    frequency++;
                    if (frequency > TangoConstants.INSTANCE.getREVIEW_FREQUENCY()) {
                        frequency = TangoConstants.INSTANCE.getREVIEW_FREQUENCY();
                    }
                }
            }

            TangoEntityManager.INSTANCE.updateData(tango.getId(),
                    "Score=" + (tango.getScore() + TangoConstants.INSTANCE.getFORGET_SCORE() / 2),
                    "Frequency=" + frequency);
            EventBus.getDefault().post(new OperateTangoEvent());
        }
    }

    public void cancelOperate() {
        if (prevTango != null && prevTango.getId() > 0) {
            TangoEntityManager.INSTANCE.updateData(prevTango.getId(),
                    "Score=" + prevTango.getScore(),
                    "Frequency=" + prevTango.getFrequency(),
                    "LastTime=" + prevTango.getLastTime().getTime());

            study = prevStudy;
            review = prevReview;
        }
    }
}

File: app/src/main/java/com/xmx/tango/module/operate/TangoOperator.kt
package com.xmx.tango.module.operate

import com.xmx.tango.core.Constants
import com.xmx.tango.common.data.DataManager
import com.xmx.tango.module.calendar.DateData
import com.xmx.tango.module.calendar.DateDataEntityManager
import com.xmx.tango.module.tango.Tango
import com.xmx.tango.module.tango.TangoConstants
import com.xmx.tango.module.tango.TangoEntityManager

import org.greenrobot.eventbus.EventBus
import java.util.*

/**
 * Created by The_onE on 2016/10/7./
 * 单语操作管理器
 */
object TangoOperator {
    var study: Int = 0 // 今日学习数
    var review: Int = 0 // 今日复习数
    private var todayConsecutive = 0 // 同一复习频率单语连续复习数

    // 保存上一次时操作的信息
    private var prevTango: Tango? = null
    private var prevStudy: Int = 0
    private var prevReview: Int = 0

    init {
        study = DataManager.getInstance().tangoStudy
        review = DataManager.getInstance().tangoReview
        // 上次运行时间
        val last = Date(DataManager.getInstance().resetLastTime)
        val lastCalendar = Calendar.getInstance()
        lastCalendar.time = last
        // 当前时间
        val now = Date()
        val nowCalendar = Calendar.getInstance()
        nowCalendar.time = now
        if (!Constants.isSameDate(now, last)) {
            if (last.time > 0) {
                // 更新上次签到日数据
                var dateData = DateDataEntityManager
                        .selectLatest("addTime", false,
                                "Year=" + (lastCalendar.get(Calendar.YEAR)),
                                "Month=" + (lastCalendar.get(Calendar.MONTH) + 1),
                                "Date=" + lastCalendar.get(Calendar.DATE))
                if (dateData != null) {
                    // 更新上次签到学习、复习数
                    DateDataEntityManager.updateData(dateData.id,
                            "Study=" + study,
                            "Review=" + review)
                } else {
                    // 创建签到记录
                    dateData = DateData()
                    dateData.year = lastCalendar.get(Calendar.YEAR)
                    dateData.month = lastCalendar.get(Calendar.MONTH) + 1
                    dateData.date = lastCalendar.get(Calendar.DATE)
                    dateData.checkIn = 1
                    dateData.study = study
                    dateData.review = review
                    dateData.addTime = now
                    DateDataEntityManager.insertData(dateData)
                }
            }
            // 今天打卡签到
            var todayData = DateDataEntityManager
                    .selectLatest("addTime", false,
                            "Year=" + (nowCalendar.get(Calendar.YEAR)),
                            "Month=" + (nowCalendar.get(Calendar.MONTH) + 1),
                            "Date=" + nowCalendar.get(Calendar.DATE))
            if (todayData == null) {
                todayData = DateData()
                todayData.year = nowCalendar.get(Calendar.YEAR)
                todayData.month = nowCalendar.get(Calendar.MONTH) + 1
                todayData.date = nowCalendar.get(Calendar.DATE)
                todayData.checkIn = 1
                todayData.addTime = now
                DateDataEntityManager.insertData(todayData)
            }
            // 清空上次数据
            study = 0
            DataManager.getInstance().tangoStudy = 0
            review = 0
            DataManager.getInstance().tangoReview = 0
            DataManager.getInstance().reviewFrequency = TangoConstants.REVIEW_FREQUENCY
            DataManager.getInstance().resetLastTime = now.time
        }
    }

    /**
     * 记住了单语
     * @param tango 记住的单语
     */
    fun remember(tango: Tango?) {
        if (tango != null && tango.id > 0) {
            // 保存操作前的状态
            prevTango = tango
            prevStudy = study
            prevReview = review

            val last = tango.lastTime
            val now = Date()
            var frequency = tango.frequency
            val goal = DataManager.getInstance().tangoGoal
            if (!Constants.isSameDate(now, last)) {
                // 今天第一次学习或复习该单语
                todayConsecutive = 0 // 重置同一复习频率单语连续复习数
                if (last != null && last.time > 0) {
                    // 复习
                    review++
                    DataManager.getInstance().tangoReview = review
                    // 每天第一次复习后降低复习频率
                    if (frequency > 0) {
                        frequency--
                    }
                } else {
                    // 学习
                    study++
                    DataManager.getInstance().tangoStudy = study
                    frequency = TangoConstants.REVIEW_FREQUENCY
                }
            } else if (study >= goal) {
                // 学习数已达到目标（在复习阶段）
                todayConsecutive++
                // 同一复习频率单语连续复习数过多
                if (todayConsecutive > TangoConstants.TODAY_CONSECUTIVE_REVIEW_MAX) {
                    todayConsecutive = 0
                    var frequencyMax = DataManager.getInstance().reviewFrequency
                    frequencyMax--
                    DataManager.getInstance().reviewFrequency = frequencyMax
                }
            }
            // 处理记住了操作的分数
            var score = TangoConstants.REMEMBER_SCORE - (study + review) / TangoConstants.TIRED_COEFFICIENT
            score = Math.max(score, TangoConstants.REMEMBER_MIN_SCORE)
            // 更新数据库中单语学习信息
            TangoEntityManager.updateData(tango.id,
                    "Score=" + (tango.score + score),
                    "Frequency=" + frequency,
                    "LastTime=" + Date().time)
            EventBus.getDefault().post(OperateTangoEvent())
        }
    }

    /**
     * 没记住操作
     * @param tango 没记住的单语
     */
    fun forget(tango: Tango?) {
        if (tango != null && tango.id > 0) {
            // 保存操作前的状态
            prevTango = tango
            prevStudy = study
            prevReview = review

            val last = tango.lastTime
            val now = Date()
            var frequency = tango.frequency
            if (!Constants.isSameDate(now, last)) {
                if (last != null && last.time > 0) {
                    // 遗忘了待复习的单语则提高复习频率
                    frequency++
                    if (frequency > TangoConstants.REVIEW_FREQUENCY) {
                        frequency = TangoConstants.REVIEW_FREQUENCY
                    }
                }
            }
            // 更新数据库中单语学习信息
            TangoEntityManager.updateData(tango.id,
                    "Score=" + (tango.score + TangoConstants.FORGET_SCORE),
                    "Frequency=" + frequency)
            EventBus.getDefault().post(OperateTangoEvent())
        }
    }

    /**
     * 彻底记住单语
     * @param tango 彻底记住的单语
     */
    fun rememberForever(tango: Tango?) {
        if (tango != null && tango.id > 0) {
            // 保存操作前的状态
            prevTango = tango
            prevStudy = study
            prevReview = review

            val last = tango.lastTime
            val now = Date()
            if (!Constants.isSameDate(now, last)) {
                if (last!!.time > 0) {
                    // 复习
                    review++
                    DataManager.getInstance().tangoReview = review
                } else {
                    // 学习
                    study++
                    DataManager.getInstance().tangoStudy = study
                }
            }
            // 复习频率为-1不会再被复习到
            val frequency = -1
            // 更新数据库中单语学习信息
            TangoEntityManager.updateData(tango.id,
                    "Score=" + (tango.score + TangoConstants.REMEMBER_FOREVER_SCORE),
                    "Frequency=" + frequency,
                    "LastTime=" + Date().time)
            EventBus.getDefault().post(OperateTangoEvent())
        }
    }

    /**
     * 未经提示拼写正确
     * @param tango 未经提示拼写正确的单语
     */
    fun rightWithoutHint(tango: Tango?) {
        if (tango != null && tango.id > 0) {
            // 保存操作前的状态
            prevTango = tango
            prevStudy = study
            prevReview = review

            val last = tango.lastTime
            val now = Date()
            var frequency = tango.frequency
            if (!Constants.isSameDate(now, last)) {
                // 今天第一次复习该单语
                if (last != null && last.time > 0) {
                    // 复习
                    review++
                    DataManager.getInstance().tangoReview = review
                }
                if (frequency > 0) {
                    // 降低复习频率
                    frequency -= 2
                    if (frequency < 0) {
                        frequency = 0
                    }
                }
            } else {
                todayConsecutive++
                // 同一复习频率单语连续复习数过多
                if (todayConsecutive > TangoConstants.TODAY_CONSECUTIVE_REVIEW_MAX) {
                    todayConsecutive = 0
                    var frequencyMax = DataManager.getInstance().reviewFrequency
                    frequencyMax--
                    DataManager.getInstance().reviewFrequency = frequencyMax
                }
            }
            // 更新数据库中单语学习信息
            TangoEntityManager.updateData(tango.id,
                    "Score=" + (tango.score + TangoConstants.REMEMBER_SCORE * 2),
                    "Frequency=" + frequency,
                    "LastTime=" + Date().time)
            EventBus.getDefault().post(OperateTangoEvent())
        }
    }

    /**
     * 经提示拼写正确
     * @param tango 经提示拼写正确的单语
     */
    fun rightWithHint(tango: Tango?) {
        if (tango != null && tango.id > 0) {
            // 保存操作前的状态
            prevTango = tango
            prevStudy = study
            prevReview = review

            val last = tango.lastTime
            val now = Date()
            var frequency = tango.frequency
            if (!Constants.isSameDate(now, last)) {
                // 今天第一次复习该单语
                if (last != null && last.time > 0) {
                    // 复习
                    review++
                    DataManager.getInstance().tangoReview = review
                }
                // 降低复习频率
                if (frequency > 0) {
                    frequency--
                }
            }
            // 更新数据库中单语学习信息
            TangoEntityManager.updateData(tango.id,
                    "Score=" + (tango.score + TangoConstants.REMEMBER_SCORE),
                    "Frequency=" + frequency,
                    "LastTime=" + Date().time)
            EventBus.getDefault().post(OperateTangoEvent())
        }
    }

    /**
     * 未拼写成功，点击查看答案
     */
    fun wrong(tango: Tango?) {
        if (tango != null && tango.id > 0) {
            // 保存操作前的状态
            prevTango = tango
            prevStudy = study
            prevReview = review

            val last = tango.lastTime
            val now = Date()
            var frequency = tango.frequency
            if (!Constants.isSameDate(now, last)) {
                // 今天第一次复习该单语
                if (last != null && last.time > 0) {
                    // 增加复习频率
                    frequency++
                    if (frequency > TangoConstants.REVIEW_FREQUENCY) {
                        frequency = TangoConstants.REVIEW_FREQUENCY
                    }
                }
            }
            // 更新数据库中单语学习信息
            TangoEntityManager.updateData(tango.id,
                    "Score=" + (tango.score + TangoConstants.FORGET_SCORE / 2),
                    "Frequency=" + frequency)
            EventBus.getDefault().post(OperateTangoEvent())
        }
    }

    /**
     * 撤销上次操作
     */
    fun cancelOperate() {
        prevTango?.apply {
            if (this.id > 0) {
                // 恢复上次保存的状态
                TangoEntityManager.updateData(this.id,
                        "Score=" + this.score,
                        "Frequency=" + this.frequency,
                        "LastTime=" + this.lastTime!!.time)
                study = prevStudy
                review = prevReview
            }
        }
    }
}
